### ISSUE TITLE
Desktop Notifications

### ISSUE THREAD

**Issue by mmmeff** [jiahaog/nativefier#88]
Heyo,

What's the effort involved in getting the desktop notifications API lined up to pass through toasts to the host OS? 

Would love to finally replace the janky chrome hangouts app once and for all.

**Comment by AdrianoCahete** [comment]
+1

**Comment by jiahaog** [comment]
Hi, what is the specific issue you're having with the notification? I believe clicking the notification will bring you back to the app if it is minimised?

**Comment by mmmeff** [comment]
I'm not getting any notifications whatsoever and assumed that meant they were unsupported, is that not the case? How are they supported currently?

**Comment by jiahaog** [comment]
Electron just supports them out of the box! If you're using facebook messenger you have to go into the messenger settings to turn on the desktop notifications. 

Could you provide me with the following:

> Version of Nativefier (run $ nativefier --version)
Version of Node.js (run $ node --version)
Command line parameters
OS and architecture you are running Nativefier from
Stack trace from the error message (if any)
Instructions to reproduce the issue

**Comment by circleride** [comment]
Same Problem, cant show notifications...
Example:

<script>
function notifyMe() {
  // Let's check if the browser supports notifications
  if (!("Notification" in window)) {
    alert("This browser does not support desktop notification");
  }

  // Let's check whether notification permissions have already been granted
  else if (Notification.permission === "granted") {
    // If it's okay let's create a notification
    var notification = new Notification("Hi there!");
  }

  // Otherwise, we need to ask the user for permission
  else if (Notification.permission !== 'denied') {
    Notification.requestPermission(function (permission) {
      // If the user accepts, let's create a notification
      if (permission === "granted") {
        var notification = new Notification("Hi there!");
      }
    });
  }

  // At last, if the user has denied notifications, and you
  // want to be respectful there is no need to bother them any more.
}Notification.requestPermission();function spawnNotification(theBody,theIcon,theTitle) {
  var options = {
      body: theBody,
      icon: theIcon
  }
  var n = new Notification(theTitle,options);
}
</script>
<button onclick="notifyMe()">Notify me!</button>

**Comment by dustinlocke** [comment]
Same problem here. No desktop notifications (Inbox & Pivotal Tracker) and verified that they're turned on from within the app.

**Comment by jiahaog** [comment]
Are you sure you have the latest version of nativefier? I've tried to nativefy http://www.bennish.net/web-notifications.html, whatsapp web, facebook messenger and it appears to work. Could you check to see if this is specific to the site you are using?

**Comment by circleride** [comment]
Not Works ;( 

![no_notifications](https://cloud.githubusercontent.com/assets/5719023/13190361/65cebabc-d722-11e5-8073-b3d018c95db5.png)

**Comment by felipead** [comment]
Same issue for me. I just installed nativefier 6.10.1 using npm 5.5.0 on latest Mac OS X El Capitan. I've got no notifications for http://hangouts.google.com.

I haven't tried other websites.

**Comment by jiahaog** [comment]
@felipead Could you try it on http://www.bennish.net/web-notifications.html

It appears to work for me on windows 10 and OSX11.2 as well...

**Comment by circleride** [comment]
Not works on Windows 8 :(

**Comment by tseven** [comment]
Notifications are not working for "hangouts.google.com" but work for "http://www.bennish.net/web-notifications.html"

OS X El Capitan: 10.11.5
Nodejs: v5.4.0
Nativefier: 7.0.1
Architecture: x64
Electron v1.1.3

I realized I'm running an older version of Node so I completely reinstalled Node to get to v6.2.2.
Same results.

I hear the notification sound for the hangouts app, but no alert popup.

**Comment by jiahaog** [comment]
@tseven I don't really use Hangouts, does it send a notification if we access it from chrome directly? Or are there certain settings you have to turn on for the desktop notifications, I believe this is the case for messenger.com

**Comment by eiger8** [comment]
Alternatively you can use [https://github.com/yakyak/yakyak](url) The same Electron build especially for hangouts.

**Comment by Mosho1** [comment]
I'm not getting notifications from gmail

**Comment by skoshy** [comment]
The http://www.bennish.net/web-notifications.html isn't showing notifications on Windows 8 for me when Nativefiering it. It says notification permission has been granted, but neither showing it immediately nor showing it after 5 seconds does anything. Works fine in Chrome.

**Comment by JiDoan** [comment]
I'm having trouble with getting any kind of mentions working for Windows 10 on the https://www.bennish.net/web-notifications.html site. Could someone share his/her configuration, command line arguments, and any changes to Windows (if any)? Attached is my configuration.

```
Version of Nativefier: 7.5.4
Version of Node.js: v6.11.2
Command line parameters: nativefier https://www.bennish.net/web-notifications.html 
OS and architecture you are running Nativefier from: Windows 10
Stack trace from the error message (if any): None
Instructions to reproduce the issue: Build then try to trigger a notification
```

**Comment by throll** [comment]
Exactly the same issue as @JiDoan on Windows 10 with nativefier 7.5.4 but works fine on Windows 7... :(

**Comment by JiDoan** [comment]
I figured out how to get notifications to work again on Windows 10 (Build 1709) thanks to the help in the [electron repo](https://github.com/electron/electron/issues/10864#issuecomment-373210477). In a nutshell, you need to register you Nativefier "app" for notifications using the Windows API ([sample code](https://emoacht.wordpress.com/2012/11/14/csharp-appusermodelid/)). Once you have your app registered, you need to add `_electron.app.setAppUserModelId("<Your app name>");` in your `resources\app\lib\main.js` file and things will work. 

Perhaps Nativefier can add a flag for Windows App ID so you don't have to edit the `main.js` file yourself?

**Comment by karnauskas** [comment]
Just tried on latest MacOSX with nativefier v7.6.1 - it doesn't allow to enable push notifications.
I used this page as an example: https://gauntface.github.io/simple-push-demo/

**Comment by jeffreysmall** [comment]
@JiDoan On windows 10, installing an app with Inno Setup. In the setup script, just make sure the AppUserModelID matches what gets put in `resources\app\lib\main.js`. That worked like a charm for me. to boot, Inno Setup is free.

**Comment by mentofa123** [comment]
@jidoan i cant get this to work. Maybe you can describe what you did exactly ._.

**Comment by rdcvnh** [comment]
Hello all,
I have created a fork which implements the **--app-user-model-id** flag to Nativefier. using @mentofa123 and @JiDoan solutions. I have also created the app ShorcutCreator.exe to create the shortcut with the AppUserModelId parameter. The only thing missing is integrating this exe with the nativefying process of an app.
The fork and more details can be found at: https://github.com/rdcvnh/nativefier

@mentofa123 and @JiDoan , I can add your names to the Credits section. Let me know if it's OK for you.

**Comment by JiDoan** [comment]
Sure I don't mind @rdcvnh, but I wouldn't have figured it out without @NicolasConstant's post in the electron repo.

**Comment by JacobDB** [comment]
@rdcvnh any plans to merge that in to the main repository?

**Comment by rdcvnh** [comment]
@JacobDB I would like to first have the ShortcutCreator.exe integrated into the build process as described on my README.md (I haven't figured out how to do so). Otherwise I think it wouldn't make sense to submit a solution that still requires some manual work.

**Comment by pseudonymcp** [comment]
@rdcvnh Maybe this could be helpful to figure it out https://github.com/KochiyaOcean/windows-shortcuts

**Comment by rdcvnh** [comment]
@pseudonymcp It seems promising. I will take a look at it. Thank you!

**Comment by artistro08** [comment]
I got the notifications to work with @rdcvnh's method. I just want them to be able to persist and when the icon is in the system tray, have that notification open the app

is there a way to do that?

**Comment by awerlang** [comment]
@artistro08 maybe this PR #640?

**Comment by haseebahmed7** [comment]
Is this fix in the final release? I just created a Skype native and am getting no notifications.

**Comment by rdcvnh** [comment]
Hi all. Sorry, I haven't had much time to look into it. If someone is interested in picking up where I left, you're more than welcome:
https://github.com/rdcvnh/nativefier

The fork should be a bit outdated by now, but the additional code should be portable to the newer versions of nativefier.
We could also get the attention of the contributors to get this implemented in original code.

**Comment by pgrenaud** [comment]
Hopefully this will get resolved soon™️ with #955. 🤞
